### PR TITLE
Add routing middleware

### PR DESCRIPTION
**Pull request by RobbeSneyders**
Part of https://github.com/spec-first/connexion/issues/1489

This PR adds a routing middleware, which adds the connexion operation to the scope passed to the next middlewares.

The middleware is implemented with a starlette router on which the api endpoints are registered. For all other routes, the router acts ass a pass-through.

### CODE REVIEW

**Comment by Ruwann** [comment]
Some more context forpossible issues of subclassing `starlette.BaseHTTPMiddleware`: https://github.com/encode/starlette/issues/1029

**Comment by RobbeSneyders** [comment]
Thanks for the additional context @Ruwann, I was indeed already running into some of these issues.
I was able to refactor and simplify the approach of the routing middleware, and I removed our own BaseHTTPMiddleware.

Note that there's some quick fixes in this PR that I still want to tackle later. Right now, I'm trying to extract all functionality into middleware with minimal effort via maximal reuse of existing connexion code (like the Api and Operation classes), and  maximal usage of Starlette tools.

File: tests/api/test_errors.py [hunk 1]
```diff
@@ -27,7 +27,6 @@ def test_errors(problem_app):
     error405 = json.loads(get_greeting.data.decode('utf-8', 'replace'))
     assert error405['type'] == 'about:blank'
     assert error405['title'] == 'Method Not Allowed'
-    assert error405['detail'] == 'The method is not allowed for the requested URL.'
```
Comment (RobbeSneyders) [inline tests/api/test_errors.py:30]: This is Flask specific and not set by Starlette.

File: tests/fakeapi/auth.py [hunk 2]
```diff
@@ -13,15 +13,3 @@ def fake_json_auth(token, required_scopes=None):
         return json.loads(token)
     except ValueError:
         return None
-
-
-async def async_basic_auth(username, password, required_scopes=None, request=None):
```
Comment (RobbeSneyders) [inline tests/fakeapi/auth.py:18]: Just some cleanup, these functions are not actually used in the tests.

**Review by RobbeSneyders — COMMENTED** [review]
This one is ready for review @Ruwann

File: connexion/middleware/routing.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    async def default_fn(self, scope: Scope, receive: Receive, send: Send) -> None:
+        """Callback to call next app as default when no matching route is found."""
+        original_scope, *_ = _scope_receive_send.get()
+
+        api_base_path = scope.get('root_path', '')[len(original_scope.get('root_path', '')):]
+
+        if CONNEXION_CONTEXT not in original_scope:
+            original_scope[CONNEXION_CONTEXT] = {}
+
+        original_scope[CONNEXION_CONTEXT].update({
```
Comment (Ruwann) [inline connexion/middleware/routing.py:89]: FYI, the `dict.setdefault()` simplifies this a bit (although perhaps a bit less untuitive as it's less known):
```
original_scope.setdefault(CONNEXION_CONTEXT, {}).update({'api_base_path': api_base_path})
```

File: connexion/middleware/routing.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            raise NotFoundProblem
+
+    async def default_fn(self, scope: Scope, receive: Receive, send: Send) -> None:
+        """Callback to call next app as default when no matching route is found."""
+        original_scope, *_ = _scope_receive_send.get()
+
+        api_base_path = scope.get('root_path', '')[len(original_scope.get('root_path', '')):]
+
+        if CONNEXION_CONTEXT not in original_scope:
+            original_scope[CONNEXION_CONTEXT] = {}
```
[the remark this answers is not in the document]
Comment (RobbeSneyders) [reply connexion/middleware/routing.py:87]: Didn't know about this. Makes sense.

File: connexion/apps/abstract.py [hunk 5]
```diff
@@ -54,8 +55,11 @@ def __init__(self, import_name, api_cls, port=None, specification_dir='',
 
         self.server = server
         self.server_args = dict() if server_args is None else server_args
+
         self.app = self.create_app()
-        self.middleware = self._apply_middleware()
+
+        middlewares = middlewares or ConnexionMiddleware.default_middlewares
```
[the remark this answers is not in the document]
Comment (RobbeSneyders) [reply connexion/apps/abstract.py:61]: Don't really see a use case for it, but the current code can indeed at least lead to some unexpected behavior. I'll put it as default for the argument or handle the `[]` explicitly.

File: connexion/apis/abstract.py [hunk 6]
```diff
@@ -109,7 +109,106 @@ def add_swagger_ui(self):
...
+    ) -> None:
+        """Minimal interface of an API, with only functionality related to routing.
+        :param specification: OpenAPI specification. Can be provided either as dict, or as path
+        to file.
+        :param base_path: Base path to host the API.
+        :param arguments: Jinja arguments to resolve in specification.
+        :param resolver: Callable that maps operationID to a function
+        :param resolver_error_handler: Callable that generates an Operation used for handling
+        ResolveErrors
+        :param debug: Flag to run in debug mode
```
[the remark this answers is not in the document]
Comment (RobbeSneyders) [reply connexion/apis/abstract.py:131]: Indeed, wanted to describe parameters of both this class and superclass, but seems like I forgot to update some. Will clean it up.

File: connexion/middleware/routing.py (hunk 3 above)
Comment (RobbeSneyders) [reply connexion/middleware/routing.py:89]: Not a fan of the one-liner, but can indeed use `setdefault` for the first part.

**Comment by RobbeSneyders** [comment]
Is this one good to go @Ruwann?

File: tests/test_middleware.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    return build_app_from_fixture('simple', request.param, middlewares=middlewares)
+
+
+def test_routing_middleware(middleware_app):
+    app_client = middleware_app.app.test_client()
+
+    response = app_client.post("/v1.0/greeting/robbe")
+
+    assert response.headers.get('operation_id') == 'fakeapi.hello.post_greeting', \
+        response.status_code
```
Comment (Ruwann) [inline tests/test_middleware.py:44]: Why the additional `response.status_code` here? Should this be a separate `assert`?

**Review by Ruwann — APPROVED** [review]

File: tests/test_middleware.py (hunk 7 above)
Comment (RobbeSneyders) [reply tests/test_middleware.py:44]: It just raises the status code as info if the test fails.

**Pull request merged**